🚨🚨 `GPT2Model` StaticCache support

poedator opened the pull request:
I copied _update_causal_mask() and _prepare_4d_causal_attention_mask_with_cache_position() from `LlamaModel`

**some tests are still failing:**

1. tests/models/gpt2/test_modeling_gpt2.py::GPT2ModelTest::test_custom_4d_attention_mask
2. test_modeling_vision_encoder_decoder.py::VIT2GPT2Test::test_save_and_load_from_pretrained

both may be linked to attention implementations. So far I was enable to figure out the reasons for failures. I'd appreciate advice or help from the maintainers.

cc: @gante

poedator committed:
initial GPT2 changes

poedator committed:
causal_mask support

poedator committed:
return_legacy_cache

poedator committed:
cleanup

poedator committed:
fix1

poedator committed:
outputs shape fixes

poedator committed:
gpt2 return fix

poedator committed:
pkv, attn fixes

poedator committed:
fix dual_head

poedator committed:
is_causal arg fix

poedator committed:
decision transformer updated

poedator committed:
style fix

poedator committed:
batch_size from inputs_embeds

poedator committed:
DecisionTransformerModel fixes

poedator committed:
cross-attn support + cache warning

poedator committed:
x-attn @decision

poedator committed:
EDCache proper init

ArthurZucker commented on src/transformers/models/decision_transformer/modeling_decision_transformer.py:
```diff
@@ -260,16 +262,17 @@ def _upcast_and_reordered_attn(self, query, key, value, attention_mask=None, hea
         self,
         hidden_states: Optional[Tuple[torch.FloatTensor]],
-        layer_past: Optional[Tuple[torch.Tensor]] = None,
+        past_key_value: Optional[Cache] = None,
+        cache_position: Optional[torch.LongTensor] = None,
```
the big issue with this is that we are breking backward compatibility for people who use `layer_past`. We need to deprecate `layer_past`!

poedator replied:
added `@deprecate_kwarg` decorator to this forward. Noted that it is an inner model class for attention, and not affecting the external model interface.

gante replied:
> Noted that it is an inner model class for attention or inner block, and not affecting the external model interface.

We agree in theory, but in practice we've broken many projects by changing the name of internal attributes and arguments. Better safe than sorry 🤗

poedator replied:
I assume that @deprecate_kwarg is sufficient here. 
 @ArthurZucker ,  Let me know if anything else is required here.

ArthurZucker commented on src/transformers/models/gpt2/modeling_gpt2.py:
```diff
@@ -806,51 +827,63 @@ def forward(
-        if position_ids is None:
-            position_ids = torch.arange(past_length, input_shape[-1] + past_length, dtype=torch.long, device=device)
-            position_ids = position_ids.unsqueeze(0)
+        # based on pattern from src/transformers/models/whisper/modeling_whisper.py::WhisperDecoder
+        return_legacy_cache = False
+        if use_cache:
+            if past_key_values is not None:
+                if isinstance(past_key_values, Cache):
+                    if self.config.add_cross_attention and not isinstance(past_key_values, EncoderDecoderCache):
+                        past_key_values = EncoderDecoderCache(past_key_values, DynamicCache())
+                elif not isinstance(past_key_values, Cache):
+                    return_legacy_cache = True
+                    logger.warning_once(
+                        "Passing a tuple of `past_key_values` is deprecated and will be removed in Transformers v4.49.0. "
+                        "You should pass an instance of `Cache` instead, e.g. "
+                        "`past_key_values=DynamicCache.from_legacy_cache(past_key_values)`."
+                    )
+                    if self.config.add_cross_attention:
+                        past_key_values = EncoderDecoderCache.from_legacy_cache(past_key_values)
+                    else:
+                        past_key_values = DynamicCache.from_legacy_cache(past_key_values)
+            elif past_key_values is None:
+                return_legacy_cache = True
+                logger.warning_once(
+                    "Passing `use_cache=True` and `past_key_values=None` will is produce cache output in legacy format.  "
+                    "This behavior is deprecated and will be changed in Transformers v4.49.0. "
+                    "To obtain output past_key_values as `Cache` instance you should pass an instance of `Cache` instead, e.g. "
+                    "`past_key_values=DynamicCache.from_legacy_cache(past_key_values)`."
+                )
+                if self.config.add_cross_attention:
+                    past_key_values = EncoderDecoderCache(DynamicCache(), DynamicCache())
+                else:
+                    past_key_values = DynamicCache()
```
From the look of it, we are adding quite a complex code, which I am not super fan of. 
Let's go with this for now, but would be nice to have a single warning to just say the one or the other is deprecated. This as is is not super readable and you have too many code pathes, when you should have: 
past_key value is None -> create DynamicCache
past_key_value is not None -> convert to Dynamic cache (not even sure that cross attention cache was even supported)
add_cross_attention -> create EncodeDDecoderCache with past_key_value and a new dynamic cache

poedator replied:
I simplified this logic according to our outline

ArthurZucker reviewed (commented):
Thanks for the PR! 
Not entirely sure it's worth adding as GPT2 is a super small model, not super optimized anymore, and fairly old so the amount of work is a bit high... 

Let's make sure we test cross attetnion path with kv cache as I am not even sure it was supported before

poedator committed:
simplified logic in `if use_cache:` for GPT2Model

poedator committed:
@deprecate_kwarg for DecisionTr attn fwd

poedator committed:
@deprecate_kwarg in gpt2

poedator commented:
> Not entirely sure it's worth adding ...
I agree in principle, but my friends use it for [Tortoise text-to-speech ](https://github.com/neonbjb/tortoise-tts) and intend to compile it (with modifications for static shapes) to accelerate.

> Let's make sure we test cross attention path with kv cache as I am not even sure it was supported before

I made effort to patch the cross-attention parts of the code as well. The relevant tests seem to pass

poedator commented:
@ArthurZucker , could you, please, approve the remaining checks workflows ?

poedator commented:
@Rocketknight1, could you kindly give feedback on this PR please

jiqing-feng commented:
Hi @poedator . Nice patch! I am trying to enable it on opt model and find you may also need to change the `_reorder_cache` function to enable beam search.

poedator commented:
> Hi @poedator . Nice patch! I am trying to enable it on opt model by following your codes if you don't mind!

hi, @jiqing-feng, Good luck in updating OPT! It should be pretty straightforward. Also look at new Cache class -related code is done for Llama. (1) this is where I borrowed from  (2) it may be closer to OPT than gpt2.

ArthurZucker commented:
sorry @poedator having a look!

gante reviewed (approved):
If the slow tests for BERT are green, LGTM 🤗 

(otherwise, feel free to ping me on slack so we can quickly sort them)

gante commented:
run slow: bert

gante commented:
(the comment above should have triggered slow tests in our CI, testing this workflow :) )

poedator commented:
this is a friendly ping to @Rocketknight1
now that @gante approved, could you finalise you review, please?

Rocketknight1 commented:
Hey! I'm not actually the core maintainer, but if @gante is happy then pinging @arthurzucker @cyrilvallez for core maintainer review

Cyrilvallez commented:
I'll have a look this afternoon!

Cyrilvallez commented on src/transformers/models/gpt2/modeling_gpt2.py:
```diff
@@ -768,7 +791,8 @@ def _prune_heads(self, heads_to_prune):
         self,
         input_ids: Optional[torch.LongTensor] = None,
-        past_key_values: Optional[Tuple[Tuple[torch.Tensor]]] = None,
+        past_key_values: Optional[Union[Tuple[Tuple[torch.Tensor]], Cache]] = None,
+        cache_position: Optional[torch.LongTensor] = None,
```
Here I'd only accept Cache format

Cyrilvallez commented on src/transformers/models/gpt2/modeling_gpt2.py:
```diff
@@ -806,51 +830,56 @@ def forward(
-                attention_mask = attention_mask.to(dtype=self.dtype)  # fp16 compatibility
-                attention_mask = (1.0 - attention_mask) * torch.finfo(self.dtype).min
+        # ._update_causal_mask() and ._prepare_4d_causal_attention_mask_with_cache_position() copied from LlamaModel
+        attention_mask = (
+            attention_mask.view(batch_size, -1) if attention_mask is not None and attention_mask.ndim < 4 else None
+        )
```
Here, if someone passes a 4d mask it's going to evaluate to `None` and later be overwritten!! If you want to allow passing 4d mask as well in this PR, this is wrong!

Cyrilvallez commented on src/transformers/models/gpt2/modeling_gpt2.py:
```diff
@@ -911,7 +935,8 @@ def forward(
                     hidden_states,
                     None,
-                    attention_mask,
+                    None,
+                    causal_mask,
```
Here i'd rather you pass along the `cache_positions` even if it is not used, for clarity

poedator replied:
converted to kwargs in a890899254d62088ac4c6b23b220e5ed8ca24ced for clarity

poedator replied:
passing as kwargs in a890899 for readability

Cyrilvallez replied:
Nop, unfortunately keyword arguments will not work with checkpointing! So they should still be passed as before as simple args, but just send `(past_key_values, cache_positions` instead of `None, None` 🤗

poedator replied:
fixed, as args

Cyrilvallez commented on src/transformers/models/gpt2/modeling_gpt2.py:
```diff
@@ -952,21 +976,145 @@ def forward(
         if output_hidden_states:
             all_hidden_states = all_hidden_states + (hidden_states,)
 
+        next_cache = past_key_values if use_cache else None
+        if return_legacy_cache:
+            next_cache = past_key_values.to_legacy_cache()
```
let's keep calling them `past_key_value` here, `next_cache` does not make sense anymore

poedator replied:
done in 70b61c7cc0c68df1cff05782691eb3f56216f70f

Cyrilvallez commented on src/transformers/models/gpt2/modeling_gpt2.py:
```diff
@@ -1037,6 +1185,7 @@ def forward(
         self,
         input_ids: Optional[torch.LongTensor] = None,
         past_key_values: Optional[Tuple[Tuple[torch.Tensor]]] = None,
+        cache_position: Optional[torch.LongTensor] = None,
```
Did you make sure they will be passed along by `generate`'s `prepare_inputd_for_generation`? 

Also, you can safely remove all `_reorder_cache()` methods il all classes

poedator replied:
removed reorder_cache in e3bd969. 
TODO: rechecking generate()

Cyrilvallez replied:
You can simply remove the Copied from line for reorder_cache in clvp model 😉

poedator replied:
done

Cyrilvallez commented on src/transformers/models/gpt2/modeling_gpt2.py:
```diff
@@ -806,51 +830,56 @@ def forward(
+                )
+                past_key_values = DynamicCache.from_legacy_cache(past_key_values)
+
+            if self.config.add_cross_attention and not isinstance(past_key_values, EncoderDecoderCache):
+                past_key_values = EncoderDecoderCache(past_key_values, DynamicCache())
```
From what I can see, gpt2 used to never use cache for cross-attention, therefore never returning it. So this would need a special treatment as well when returning the kv (only return self_attention part, and in legacy format)

poedator replied:
Originally I borrowed code from modeling_whisper.py
later I simplified it on suggestion from Arthur. 
Glad to change it again, but please, give more specific instructions on how you see it and which tests it should pass

Cyrilvallez replied:
It should basically pass all the tests haha 😉 Here I just meant it needs a special treatment as well when returning the cache. So basically flag `return_legacy_cache` should be active, and when returning do: 
```py
if return_legacy_cache:
        past_key_values = past_key_values.self_attention_cache.to_legacy_cache() if add_cross_attentions else past_key_values.to_legacy_cache()
```

poedator replied:
fixed it now, all GPT2 tests pass OK

Cyrilvallez reviewed (commented):
Ok, here is my review! I only reviewed gpt2, as the other model seem to only be copy-pasted of the same changes (minus some parts), but of course the relevant comments apply as well! 🤗

Very nice work overall, congrats! Just a few nits, and an issue about the 4d masking but otherwise all good 👌

Cyrilvallez commented:
Also, did you run slow tests by any chance? 🤗

poedator committed:
deprecation version updated to 4.51

poedator commented on src/transformers/models/gpt2/modeling_gpt2.py:
```diff
@@ -299,18 +304,18 @@ def forward(
-
-        is_cross_attention = encoder_hidden_states is not None
-        is_causal = attention_mask is None and query_states.shape[-2] > 1 and not is_cross_attention
+        self.is_causal = attention_mask is None and query_states.shape[-2] > 1 and not is_cross_attention
```
indeed this is not the only way to do it, but I kept the original line from here:
https://github.com/huggingface/transformers/blob/884a8ea1f058716c24adb52a0a6a0bf41fbb973d/src/transformers/models/gpt2/modeling_gpt2.py#L313

poedator commented on src/transformers/models/gpt2/modeling_gpt2.py:
```diff
@@ -267,19 +270,21 @@ def _upcast_and_reordered_attn(self, query, key, value, attention_mask=None, hea
 
         return attn_output, attn_weights
 
+    @deprecate_kwarg("layer_past", new_name="past_key_value", version="4.49.0", raise_if_both_names=True)
```
addressed in 13725053f4ee40ef4bb977a63ff4f5988c1966d4

poedator commented on src/transformers/models/gpt2/modeling_gpt2.py:
```diff
@@ -388,10 +392,12 @@ def __init__(self, config, layer_idx=None):
 
         self.mlp = GPT2MLP(inner_dim, config)
 
+    @deprecate_kwarg("layer_past", new_name="past_key_value", version="4.49.0", raise_if_both_names=True)
```
done

poedator committed:
kwargs in gradient_checkpointing_fn

poedator committed:
rename next_cache to past_key_values

poedator commented on src/transformers/models/gpt2/modeling_gpt2.py:
```diff
@@ -340,15 +345,14 @@ def forward(
             )
 
         attn_output = attn_output.reshape(*attn_output.shape[:-2], -1).contiguous()
         attn_output = self.c_proj(attn_output)
         attn_output = self.resid_dropout(attn_output)
 
-        outputs = (attn_output, present)
+        outputs = (attn_output,)
         if output_attentions:
             outputs += (attn_weights,)
```
we may not be able to drop `output_attentions` here entirely, since it drives choice between sdpa and eager here:
https://github.com/huggingface/transformers/blob/2ab7bdc40333b230b642f09e8334fb8e1a92d2a4/src/transformers/models/gpt2/modeling_gpt2.py#L318

I'd much prefer to keep this PR focused on Cache class support, and leave additional refactoring for later.

Cyrilvallez replied:
Ha yes sorry, I meant to grab the arg from `kwargs`, e.g. `kwargs.get("output_attentions", False)`. But anyway, let's always return both, independently of the arg in the Attention: `return attn_output, attn_weights`. Then use the arg for the return only in `Block`

poedator replied:
Got you now. Moved all `output_attentions` logic to the Block

poedator commented on src/transformers/models/gpt2/modeling_gpt2.py:
```diff
@@ -403,7 +409,8 @@ def forward(
         hidden_states = self.ln_1(hidden_states)
         attn_outputs = self.attn(
             hidden_states,
-            layer_past=layer_past,
+            past_key_value=past_key_value,
+            cache_position=cache_position,
             attention_mask=attention_mask,
             head_mask=head_mask,
             use_cache=use_cache,
```
same as above, `output_attentions`  drives choice between `sdpa` and `eager`

Cyrilvallez replied:
Sure but let's return both anyway!

poedator commented on src/transformers/models/gpt2/modeling_gpt2.py:
```diff
@@ -425,6 +432,7 @@ def forward(
             hidden_states = self.ln_cross_attn(hidden_states)
             cross_attn_outputs = self.crossattention(
                 hidden_states,
+                past_key_value=past_key_value,
                 attention_mask=attention_mask,
```
same as above, need to keep `output_attentions` which makes `cross_attn_outputs` tuple size uncertain

Cyrilvallez replied:
see above as well

poedator replied:
done

poedator committed:
attention_mask prep

poedator commented:
@Cyrilvallez 
> Also, did you run slow tests by any chance? 🤗

I did run the SLOW tests, and all of them passed before I asked for the review. There were couple of exceptions in some other models' tests which did not appear in CI upon submission. 

Will run the SLOW tests again and attach the output link here before calling for another review.

I addressed most of your comments. Please see if you could resolve some or all of them, or give me more guidance

poedator committed:
+cache_position in GPT2DoubleHeadsModel

poedator committed:
undo kwargs in gradient checkpointing

poedator committed:
moved up `if self.gradient_checkpointing`

poedator committed:
consistency in decision_transformer

poedator commented:
Status as of 2025-03-01
- Addressed comments from @Cyrilvallez (see above, need resolving)
- rebased to recent main
- ran SLOW tests locally: 2 tests failed, see logs here: https://gist.github.com/poedator/f4bdaac321440dd4110af0ee89ae9905
    (1) `tests/models/gpt2/test_modeling_gpt2.py::GPT2ModelLanguageGenerationTest::test_flash_attn_2_generate_padding_left` - has different text generated with FA2. reasons unclear. Generation with eager matches fine
    (2) `tests/models/gpt2/test_modeling_gpt2.py::GPT2ModelLanguageGenerationTest::test_gpt2_sample_max_time` - longer test time, reasons unclear.
    
Also observed errors in CI:
    (3) `tests/models/vision_encoder_decoder/test_modeling_vision_encoder_decoder.py::VIT2GPT2Test::test_save_and_load_from_pretrained` - somehow loaded model uses eager, not sdpa and diverges. but difference is larger than with main. If both models use eager, generation matches perfectly.
    (4) `tests/models/vision_encoder_decoder/test_modeling_vision_encoder_decoder.py::Donut2GPT2Test::test_save_and_load_from_pretrained` - probably same issue    

calling @Cyrilvallez to resolve issues and suggest how to fix the tests problems.

Cyrilvallez commented:
Just answered on the separate comments! Let me know if something is still unclear! For the slow tests, if the particular test is not passing on main as well, it's fine! You should just make sure that this PR does not introduce any new failure!

poedator committed:
pastkv, cache_pos in grad_checkpt args

poedator committed:
rm _reorder_cache

poedator committed:
output_attentions streamlined

poedator committed:
decision_transformer consistency

poedator committed:
return_legacy_cache improved

poedator committed:
ClvpForCausalLM used for legacy cache test now

poedator committed:
is_causal fixed

poedator committed:
attn_output cleanup

poedator committed:
consistency @ decision_transformer

poedator committed:
Updated deprecation notice version to 4.52

poedator commented:
Hi,  @Cyrilvallez !
it took me some time to fix a bug, but now this PR passes all SLOW tests locally. The CI tests have only one issue related to gated Mistral mode. So could you please, run the remaining workflows and finalize the review?

jiqing-feng commented:
Looking forward to this PR merged!

BTW, I don't understand why we keep cross attention in gpt2 as I didn't see any GPT2 arch model have encoder hidden states. GPT is a decoder-only model. Could I have an example or an existing model in HF model hub which has encoder_hhidden_states and uses cross attention? If not, can we remove the cross attention code so we can just use static cache?

jiqing-feng commented:
> Looking forward to this PR merged!
> 
> BTW, I don't understand why we keep cross attention in gpt2 as I didn't see any GPT2 arch model have encoder hidden states. GPT is a decoder-only model. Could I have an example or an existing model in HF model hub which has encoder_hhidden_states and uses cross attention? If not, can we remove the cross attention code so we can just use static cache?

Hi @SunMarc @ArthurZucker @Rocketknight1 . I'd like to hear your opinion, and also want to know what need to be changed before merging this PR. The static cache is essential for torch.compile. We'd like to see gpt2 gain this optimization!

Cyrilvallez commented on src/transformers/models/decision_transformer/modeling_decision_transformer.py:
```diff
@@ -689,16 +702,20 @@ def forward(
         if output_hidden_states:
             all_hidden_states = all_hidden_states + (hidden_states,)
 
+        next_cache = past_key_values if use_cache else None
+        if return_legacy_cache:
+            next_cache = past_key_values.to_legacy_cache()
```
Let's not use a name such as `next_cache` here, which is confusing now!

poedator replied:
done, made consistent with gpt2

Cyrilvallez commented on src/transformers/models/decision_transformer/modeling_decision_transformer.py:
```diff
@@ -558,14 +561,28 @@ def forward(
+            if not isinstance(past_key_values, Cache):
+                return_legacy_cache = True
+                logger.warning_once(
+                    "Passing a tuple of `past_key_values` is deprecated and will be removed in Transformers v4.49.0. "
+                    "You should pass an instance of `Cache` instead, e.g. "
```
We should change the deprecation version here. 4.53 should do it!

Cyrilvallez commented on src/transformers/models/decision_transformer/modeling_decision_transformer.py:
```diff
@@ -383,10 +384,12 @@ def __init__(self, config, layer_idx=None):
 
         self.mlp = DecisionTransformerGPT2MLP(inner_dim, config)
 
+    @deprecate_kwarg("layer_past", new_name="past_key_value", version="4.51.0", raise_if_both_names=True)
```
Let's use `4.53` here

Cyrilvallez commented on src/transformers/models/gpt2/modeling_gpt2.py:
```diff
@@ -267,19 +270,21 @@ def _upcast_and_reordered_attn(self, query, key, value, attention_mask=None, hea
 
         return attn_output, attn_weights
 
+    @deprecate_kwarg("layer_past", new_name="past_key_value", version="4.52.0", raise_if_both_names=True)
```
similarly here for version

Cyrilvallez commented on src/transformers/models/gpt2/modeling_gpt2.py:
```diff
@@ -388,10 +389,12 @@ def __init__(self, config, layer_idx=None):
 
         self.mlp = GPT2MLP(inner_dim, config)
 
+    @deprecate_kwarg("layer_past", new_name="past_key_value", version="4.51.0", raise_if_both_names=True)
```
same

Cyrilvallez commented on src/transformers/models/gpt2/modeling_gpt2.py:
```diff
@@ -806,51 +829,62 @@ def forward(
+                    "`past_key_values=DynamicCache.from_legacy_cache(past_key_values)`."
+                )
+                past_key_values = DynamicCache.from_legacy_cache(past_key_values)
+
+            if self.config.add_cross_attention and not isinstance(past_key_values, EncoderDecoderCache):
+                past_key_values = EncoderDecoderCache(past_key_values, DynamicCache())
```
Any reason why decision_transformer is not doing the same with EncoderDecoderCache here?

poedator replied:
fixed, made consistent with gpt2

Cyrilvallez reviewed (commented):
Hey @poedator! Super super sorry, this got buried 🥲 Here is a long due review. Basically we need to update the deprecation versions mostly now 😅

Also, I see that there are differences between the 2 changed models. Are there any reasons for that? They should be closer to one-another

poedator committed:
upd deprecation

poedator committed:
consistent legacy cache code in decision transformers\

poedator committed:
next_cache -> past_kv in decision_tr

poedator commented:
> Hey @poedator! Here is a long due review. [...]

Hi, @Cyrilvallez! Hope that the merge is close. You were correct about inconsistencies with the `decision_transformer`. I fixed those, updated deprecation, rebased to fresh main. Let's see if this is good enough

Cyrilvallez commented on src/transformers/models/gpt2/modeling_gpt2.py:
```diff
@@ -906,51 +929,62 @@ def forward(
+        if use_cache:
+            if past_key_values is None:
+                return_legacy_cache = True
+                logger.warning_once(
+                    "Passing `use_cache=True` and `past_key_values=None` will is produce cache output in legacy format.  "
+                    "This behavior is deprecated and will be changed in Transformers v4.53.0. "
+                    "To obtain output past_key_values as `Cache` instance you should pass an instance of `Cache` instead, e.g. "
+                    "`past_key_values=DynamicCache.from_legacy_cache(past_key_values)`."
+                )
```
Here let's remove this warning, as it will be surprising for users and not needed (as we return the same format as before anyway). Let's only keep the other one, when KV are passed as tuple!

Cyrilvallez commented on src/transformers/models/gpt2/modeling_gpt2.py:
```diff
@@ -565,6 +567,8 @@ class GPT2PreTrainedModel(PreTrainedModel):
     _skip_keys_device_placement = "past_key_values"
     _supports_flash_attn_2 = True
     _supports_sdpa = True
+    _supports_cache_class = True
+    _supports_static_cache = True
```
Here, `decision_transformer` should have the flags set to True and the updated docstrings as well!

poedator replied:
thank you for pointing these flags to me. I made matching edits to `modeling_decision_transformer.py`
However, I am unclear about the docstrings. There is no direct equivalent for `GPT2_INPUTS_DOCSTRING` and `DECISION_TRANSFORMER_INPUTS_DOCSTRING` has quite different format.

Cyrilvallez reviewed (commented):
Hey @poedator! Thanks a lot for updating! Here are some final thoughts! Also, I still see some logic differences between both models, most notable the mask handling! Let's make sure both models are functionyla equivalent 🤗

poedator committed:
cache support flags in decision_transf

poedator committed:
rm legacy cache warning

poedator committed:
consistency in cache init for decision transf

poedator commented:
> Hey @poedator! Thanks a lot for updating! Here are some final thoughts! Also, I still see some logic differences between both models, most notable the mask handling! Let's make sure both models are functionyla equivalent 🤗

I fixed what you directly suggested and also made the cache init sections between the 2 models more similar. 

As for  mask handling and the rest of logic differences -- I am less familiar with this exotic `decision transformer` model, so  I'd be glad if you point me to the specific failing tests so I'd debug the problem. Or feel free to edit this PR directly and take 50% of the glory.
the CI tests seem to fail for unrelated reasons.

Cyrilvallez commented on src/transformers/models/decision_transformer/modeling_decision_transformer.py:
```diff
@@ -456,6 +458,8 @@ class DecisionTransformerGPT2PreTrainedModel(PreTrainedModel):
     is_parallelizable = True
     supports_gradient_checkpointing = True
+    _supports_cache_class = True
+    _supports_static_cache = True
```
Alright, then it should not support static cache!

poedator replied:
changed  it this way:
` _supports_cache_class = True`
` _supports_static_cache = True`

Cyrilvallez reviewed (commented):
Oh sorry, I thought you wanted to add the same changes and functionalities to decision_transformer as well. If it's not the case, we can just remove the flag saying that it supports static cache (as we would need correct masking for it), then it should be good to go! 🤗

poedator committed:
no Static Cache for Decision Transformer

poedator commented:
> Oh sorry, I thought you wanted to add the same changes and functionalities to decision_transformer as well. If it's not the case, we can just remove the flag saying that it supports static cache (as we would need correct masking for it), then it should be good to go! 🤗

Yes, I was focused on adding support of StaticCache to GPT2, as it is a rather popular model. I will gladly leave the last bit of the Decision Transformer challenge to whoever knows that model better. Please see if this is OK now.

Cyrilvallez committed:
Merge branch 'main' into gpt_static

Cyrilvallez reviewed (approved):
Alright, LGTM!! 🤗🚀 Thanks a lot for the addition, and once again super sorry for the time it took! Great work 🤗
All tests (slow as well) are passing so it's good to go!

poedator commented:
> Alright, LGTM!! 🤗🚀 ... good to go!

@Cyrilvallez @ArthurZucker @gante 
thank you for the valuable inputs and support along the way. Hopefully there are still quite a few people who use GPT2 and who may find this PR useful
